cfdm/data/abstract/mesharray.py

```diff
@@ -0,0 +1,366 @@
+                # When the shape is unknown, there must be (and so we
+                # are assuming that there is) only one pass through
+                # this loop.
+                u = subarray[...]
+
+        if not known_shape:
+            # Store the shape, now that is it known.
+            self._set_component("shape", u.shape, copy=False)
+
+        if indices is Ellipsis:
+            return u
+
+        return self.get_subspace(u, indices, copy=False)
+
+    @property
+    def dtype(self):
+        """The data-type of the uncompressed data.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        """
+        return self._get_compressed_Array().dtype
+
+    def get_cell_dimension(self, default=ValueError()):
+        """Return the position of the cell dimension.
+
+        In UGRID, this is provided by one of the the "edge_dimension",
+        "face_dimension" or "volume_dimension" variables.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        :Parameters:
+
+            default: optional
+                Return the value of the *default* parameter if there
+                is no cell dimension position.
+
+                {{default Exception}}
+
+        :Returns:
+
+            `int`
+                The position of the cell dimension.
+
+        """
+        return self._get_component("cell_dimension", default)
+
+    def get_start_index(self, default=ValueError()):
+        """Return the start index.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        :Parameters:
+
+            default: optional
+                Return the value of the *default* parameter if there
+                is no start index.
+
+                {{default Exception}}
+
+        :Returns:
+
+            `int`
+                The start index.
+
+        """
+        return self._get_component("start_index", default)
+
+    def subarray_parameters(self):
+        """Non-data parameters required by the `Subarray` class.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        :Returns:
+
+            `dict`
+                The parameters, with keys ``'compressed_dimensions'``,
+                ``'start_index'``, and ``'cell_dimension'``.
+
+        """
+        return {
+            **super().subarray_parameters(),
+            "start_index": self.get_start_index(),
+            "cell_dimension": self.get_cell_dimension(),
+        }
+
+    def subarray_shapes(self, shapes):
+        """Create the subarray shapes along each uncompressed dimension.
+
+        Note that the ouput is indpendent of the *shapes* parameter,
+        because each dimension of the compressed data corresponds to a
+        unique dimension of the uncompressed data.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        .. seealso:: `subarray`
+
+        :Parameters:
+
+            {{subarray_shapes chunks: `int`, sequence, `dict`, or `str`, optional}}
+
+        :Returns:
+
+            `list`
+                The subarray sizes along each uncompressed dimension.
```

Comment:
```suggestion
                The subarray shapes along each uncompressed dimension.
```

Reply:
Oh fair. I guess I am quite sensitive to `shape` and `size` being different from the NumPy context,  hence raising the comment, but didn't appreciate the subsequent "along each dimension" which does map from `size` to `shape`. SO all good, marking as resolved. Thanks.